docs: Refactor GPU architecture section (#191)
bentoml/llm-inference-in-production: docs/kernel-optimization/gpu-architecture-fundamentals/gpu-memory.md

private set of registers, meaning other threads cannot access them. Physically,
however, these registers come from a large register file on the
[SM](/kernel-optimization/gpu-architecture-fundamentals/streaming-multiprocessors/)
(for example, 256 KB per SM on H100), which is shared across all resident
threads and partitioned among them.

Register access is much cheaper than going to shared memory or HBM because the
data is already on-chip and directly available to the compute units.

Registers are a limited resource. Higher register use per thread can reduce the
number of resident threads and warps when register allocation becomes the
binding limit. This is a primary driver of the occupancy tradeoff.

## Shared memory (SMEM) and L1 cache

Modern GPUs provide two types of fast, on-chip memory inside each SM: shared
memory (SMEM) and the L1 cache. They often use the same physical SRAM, but their
roles are different.

L1 cache is hardware-managed. When a thread reads from global memory (HBM), the
GPU may store the data in L1 automatically. If the same data is accessed again,
it can be served from L1 much faster (a cache hit) instead of going back to HBM.

> This process is transparent to the programmer: you do not explicitly load data

Reviewer: ```suggestion
This process is transparent to the programmer: you don’t explicitly load data
```

Shared memory, in contrast, is programmer-managed. It is a small, explicitly
allocated memory space that is shared by all threads in a thread block. Threads
can read and write shared memory directly, and use synchronization (e.g.,
`__syncthreads()`) to coordinate access. This makes shared memory a predictable
and controllable workspace for cooperation between threads.

The main purpose of shared memory is to reduce expensive HBM accesses. A common
pattern is to load data once from global memory into shared memory, then reuse
it multiple times across threads. Because shared memory is on-chip and much
faster than HBM, this can significantly improve performance. This pattern
appears in many high-performance kernels, including
[matrix multiplication](/kernel-optimization/gpu-architecture-fundamentals/tensor-cores/)
and [attention](/kernel-optimization/flashattention/).

Shared memory is organized into banks (typically 32). When multiple threads in a
[warp](/kernel-optimization/gpu-architecture-fundamentals/threads-warps-blocks/#warps)
access different addresses in the same bank simultaneously, a bank conflict
occurs and the accesses are serialized (threads reading the same address are

> served by a broadcast and do not conflict). Avoiding bank conflicts is a common

Reviewer: ```suggestion
served by a broadcast and don’t conflict). Avoiding bank conflicts is a common
```
